File: 2023-01/B287e.py
class Trie:
    def __init__(self):
        self.d = {}
        self.tail = 0
        self.count = 0
        
    def find(self, s, i):
        if i == len(s):
            if self.count > 1:
                return i
            return i - 1
        c = s[i]
        if self.count <= 1 or c not in self.d:
            return i - 1
        return self.d[c].find(s, i+1)
    
    def add(self, s, i):
        if i == len(s):
            self.tail += 1
            return
        c = s[i]
        if c not in self.d:
            self.d[c] = Trie()
        self.d[c].add(s, i+1)
        
    def trace(self):
        self.count = self.tail
        for v in self.d.values():
            self.count += v.trace()
        return self.count
        
    def debug(self):
        if not self.d:
            return '[.]'
        ret = '['
        for k, v in self.d.items():
            ret += k + str(self.count) + v.debug()
        ret += ']'
        return ret
        
n = int(input())
ls = []
T = Trie()
for _ in range(n):
    s = input()
    T.add(s, 0)
    ls.append(s)
T.trace()
# print(T.debug())
for s in ls:
    print(T.find(s, 0))

File: archived/2023-01/B287e.cpp
#include <bits/stdc++.h>
using namespace std;

class Trie {
    unordered_map<char, Trie*> d;
    int tail = 0;
    int count = 0;
public:
    int find(string &s, int i) {
        if (i == (int) s.size()) {
            if (count > 1) return i;
            return i - 1;
        }
        char c = s[i];
        if (count <= 1 or !d.count(c)) return i - 1;
        return d[c]->find(s, i+1);
    }

    void add(string &s, int i) {
        if (i == (int) s.size()) {
            tail++;
            return;
        }
        char c = s[i];
        if (!d.count(c)) d[c] = new Trie();
        d[c]->add(s, i+1);
    }
    
    int trace() {
        count = tail;
        for (auto p: d) count += p.second->trace();
        return count;
    }
};

int main() {
    int n;
    cin >> n;
    vector<string> v;
    string s;
    Trie* T = new Trie();
    while (n--) {
        cin >> s;
        v.emplace_back(s);
        T->add(s, 0);
    }
    T->trace();
    for (string &s: v) {
        cout << T->find(s, 0) << endl;
    }
}